### PR TITLE
Reduce redundant code in CUDF JNI

### PR DESCRIPTION
**Pull request by mythrocks**
This commit cleans up some of the redundancy in CUDF JNI native code, and modifies some of the `unique_ptr` handling:

### Improvements
1. Added utility function to convert from `native_jpointerArray<column_view>` to `vector<column_view>`. Removed redundant code in `stringConcat` functions.
2. Added utility to convert from `unique_ptr<column>` to `jlong`. This eliminates variables from nearly all the JNI functions.
3. Added utility functions for extracting `std::vector` from `native_jArray` classes. This reduces raw loops in `TableJni.cpp`.
4. Eliminated unnecessary use of `unique_ptr` in locations where they are immediately `release()`d. (E.g. `ColumnVectorJni::getNativeColumnView()`.)
5. Removed unnecessary `std::move()` calls.
6. Removed unnecessary `reinterpret_cast` (E.g. in `copyColumnViewToCV()`.)
7. `typedef` -> `using`.

### CODE REVIEW

File: java/src/main/native/include/jni_utils.hpp
Comment (jlowe) [inline java/src/main/native/include/jni_utils.hpp:88]: I'm not a fan of these seemingly innocuous methods with very nasty side-effects.  They look a _lot_ like the first form, which simply does a cast, while these are also releasing the memory ownership from underneath the `std::unique_ptr`.  The forms that release the memory should have `release` in their name or some other indicator that this isn't just a casting helper.

File: java/src/main/native/include/jni_utils.hpp
Comment (jlowe) [inline java/src/main/native/include/jni_utils.hpp:15]: 2022 copyrights

File: java/src/main/native/src/ColumnViewJni.cpp
Comment (jlowe) [inline java/src/main/native/src/ColumnViewJni.cpp:75]: 2022 copyrights

File: java/src/main/native/src/ColumnViewJni.cpp
Comment (jlowe) [inline java/src/main/native/src/ColumnViewJni.cpp:1477]: Isn't this as simple as making this a `cudf::table_view` local variable rather than trying to allocate a table view on the heap?

Comment (mythrocks) [reply java/src/main/native/src/ColumnViewJni.cpp:1477]: Yep. I haven't fixed this one yet. I'm only halfway through this file, so far.

File: java/src/main/native/include/jni_utils.hpp
Comment (mythrocks) [reply java/src/main/native/include/jni_utils.hpp:88]: I'm open to using a better name for these functions. `release_to_jlong()`?

Comment (jlowe) [reply java/src/main/native/include/jni_utils.hpp:88]: That works, or maybe `release_as_jlong()`

Comment (mythrocks) [reply java/src/main/native/include/jni_utils.hpp:88]: That's much better. Will post the change shortly.

Comment (mythrocks) [reply java/src/main/native/include/jni_utils.hpp]: Fixed.

File: java/src/main/native/src/ColumnViewJni.cpp
Comment (mythrocks) [reply java/src/main/native/src/ColumnViewJni.cpp:75]: Fixed.

File: java/src/main/native/src/ColumnViewJni.cpp
Comment (mythrocks) [reply java/src/main/native/src/ColumnViewJni.cpp]: This file has now been dealt with.

**Comment by mythrocks** [comment]
Though it could stand to be improved, I think I'll leave `convert_table_for_return()` intact for this iteration.

File: java/src/main/native/include/jni_utils.hpp
Comment (revans2) [inline java/src/main/native/include/jni_utils.hpp:78]: nit: I am not thrilled with the name. Could we call this `ptr_to_jlong` or `addr_to_jlong` so it is clear that this returns the address as a `jlong` and not the value in it?

**Review by revans2 — COMMENTED** [review]
Looking really good

File: java/src/main/native/include/jni_utils.hpp
Comment (mythrocks) [reply java/src/main/native/include/jni_utils.hpp:78]: Hmm. Thanks for pointing that out, @revans2. That hadn't occurred to me. I'll make the change. 
`ptr_to_jlong()`, if that's acceptable?

File: java/src/main/native/src/ColumnVectorJni.cpp
Comment (jlowe) [inline java/src/main/native/src/ColumnVectorJni.cpp:362]: I'm not a fan of giant, refactoring PRs that do more than just their main refactor, and this is doing just that.  This is tying an important memory leak bugfix to a completely unrelated change, and it would be much more expedient to review and easier to manage going forward if this fix were committed separately.

File: java/src/main/native/include/jni_utils.hpp
Comment (revans2) [reply java/src/main/native/include/jni_utils.hpp:78]: I am fine with that.

File: java/src/main/native/src/ColumnVectorJni.cpp
Comment (revans2) [reply java/src/main/native/src/ColumnVectorJni.cpp:362]: +1

Comment (mythrocks) [reply java/src/main/native/src/ColumnVectorJni.cpp:362]: > This is tying an important memory leak bugfix

Since we hadn't noticed it before, I didn't think the memory leak was important, per se. But I do see your point. Moving those fixes to a different commit would make them a candidate for 22.02.

I will update here with a separate PR for just the bug portion of this change.

Comment (mythrocks) [reply java/src/main/native/src/ColumnVectorJni.cpp:362]: > I will update here with a separate PR...

I'm done reviewing `ColumnViewJni.cpp` and `ColumnVectorJni.cpp. 
The bug PR will have to wait until I'm done scanning through `TableJni.cpp` as well. It shouldn't be long...

Comment (mythrocks) [reply java/src/main/native/src/ColumnVectorJni.cpp:362]: I have filed #10029 for the memory leaks. I will rebase this branch once that PR is checked in.

**Comment by mythrocks** [comment]
Will rebase to `branch-22.04` shortly.

**Comment by mythrocks** [comment]
Re-targeting `branch-22.04` is causing the changes from #10026 to appear in this PR. This is likely transient, and should be corrected when #10026 finds its way to `branch-22.04`.
The material changes in this PR are only in:
1. [`jni_utils.hpp`](https://github.com/rapidsai/cudf/pull/10019/files#diff-7a32a5ca0549ea6501d99100b7c2a4b1ce1b604920f72494e3eb744b0f1890b7)
2. [`ColumnVectorJni.cpp`](https://github.com/rapidsai/cudf/pull/10019/files#diff-14cbf047b47b61d9683c809efc1337b1507ce9b1b4759aee79f745232d2bc178)
3. [`ColumnViewJni.cpp`](https://github.com/rapidsai/cudf/pull/10019/files#diff-2c9cdaa1def3de141370de960a4162392d2959a8d2a8d6462fbfdc7b207212e0)
4. [`TableJni.cpp`](https://github.com/rapidsai/cudf/pull/10019/files#diff-342d706574d8fc9c3042623eead4ebc370863d3d40e921fe78c2c09b04e93c96)

File: java/src/main/native/include/jni_utils.hpp
Comment (mythrocks) [reply java/src/main/native/include/jni_utils.hpp]: Changed to `ptr_as_jlong()`. I would have preferred the shorter name, but it's best to avoid the ambiguity that @revans2 pointed out.

**Comment by mythrocks** [comment]
I've rebased this on `branch-22.04`. Further modifications (E.g. to `convert_table_for_return()`) can be deferred to a follow-up.

**Review by jlowe — APPROVED** [review]

**Comment by mythrocks** [comment]
Rerun tests

**Comment by jlowe** [comment]
I just noticed that there are a number of `reinterpret_cast<jlong>` occurrences still in TableJni.cpp, as well as a number remaining elsewhere in other files, including ones that are of the form `reinterpret_cast<jlong>(x.release())`.  Were these left intentionally?

**Comment by mythrocks** [comment]
> I just noticed that there are a number of reinterpret_cast<jlong> occurrences still in `TableJni.cpp`...

The ones I left alone (e.g. the join code you added most recently) seemed fine to my eye, at the time. I should probably make another pass and confirm. The `reinterpret_cast<jlong>(x.release())` sounds avoidable.

**Comment by mythrocks** [comment]
> there are a number of `reinterpret_cast<jlong>` occurrences

These occurrences have been removed from `ColumnVectorJni.cpp`, `ColumnViewJni.cpp`, and `TableJni.cpp`.

**Comment by mythrocks** [comment]
Rebased to latest, including #10040. Waiting on CI before merging.

**Comment by mythrocks** [comment]
@gpucibot merge

**Comment by mythrocks** [comment]
Thanks for the reviews, @revans2, @jlowe. This is now merged.

**Pull request merged**
